### PR TITLE
NULS 2.0 integration

### PR DESCRIPTION
**Pull request by ccccbjcn**
## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [ ] Update documentation as needed.

### CODE REVIEW

File: coins.json [hunk 1]
```diff
@@ -705,6 +705,23 @@
...
+        "decimals": 8,
+        "blockchain": "NULS",
+        "derivationPath": "m/44'/8964'/0'/0/0",
+        "curve": "secp256k1",
+        "publicKeyType": "secp256k1",
+        "explorer": "https://nulscan.io/",
+        "info": {
+            "url": "https://nuls.io",
+            "client": "https://github.com/nuls-io/nuls-v2",
+            "clientPublic": "",
```
Comment (hewigovens) [inline coins.json]: could you please provide this https://developer.trustwallet.com/wallet-core/rpc-requirements?

File: src/Coin.cpp [hunk 2]
```diff
@@ -40,6 +40,10 @@
 #include "Wanchain/Address.h"
 #include "Waves/Address.h"
 #include "Zcash/TAddress.h"
+#include "NULS/Address.h"
+#include "Steem/Address.h"
+#include "EOS/Address.h"
+#include "IoTeX/Address.h"
```
Comment (hewigovens) [inline src/Coin.cpp]: duplicated imports

File: src/NULS/Address.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+uint16_t Address::chainID() const {
+    return decode16LE(bytes.data());
+}
+
+uint8_t Address::type() const {
+    return bytes[2];
+}
+
+TW::PrivateKey Address::importHexPrivateKey(const std::string& hexPrivateKey) {
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: could you remove this? we will only support import from recovery phrases

File: include/TrustWalletCore/TWNULSAddress.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+TW_EXPORT_STATIC_METHOD
+bool TWNULSAddressIsValidString(TWString *_Nonnull string);
+
+/// Creates an address from a string representaion.
+TW_EXPORT_STATIC_METHOD
+struct TWNULSAddress *_Nullable TWNULSAddressCreateWithString(TWString *_Nonnull string);
+
+/// Creates an address from a raw data representaion.
+TW_EXPORT_STATIC_METHOD
+struct TWNULSAddress *_Nullable TWNULSAddressCreateWithData(TWData *_Nonnull data);
```
Comment (hewigovens) [inline include/TrustWalletCore/TWNULSAddress.h]: can you remove this function?

File: src/NULS/Address.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+// file LICENSE at the root of the source code distribution tree.
+#include "Address.h"
+
+#include <TrezorCrypto/ecdsa.h>
+
+#include "../Base58.h"
+#include "../BinaryCoding.h"
+#include "../HexCoding.h"
+
+using namespace TW::NULS;
```
Comment (hewigovens) [inline src/NULS/Address.cpp]:
```suggestion
using namespace TW;
using namespace TW::NULS;
```
you can remove `TW::` prefix in code

File: src/NULS/Address.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    }
+    }
+
+    auto key = PrivateKey(data);
+    return key;
+}
+
+std::string Address::string() const {
+    std::string temp = prefix;
+    return temp.append(TW::Base58::bitcoin.encode(bytes.begin(), bytes.end()));
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: simplify like this?
```suggestion
    return prefix + Base58::bitcoin.encode(bytes.begin(), bytes.end());
```

File: tests/NULS/SignerTests.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+    input.set_timestamp(0x5d8885f8);
+    return input;
+}
+
+TEST(NULSSigner, Sign) {
+    auto input = makeInput("9ce21dad67e0f0af2599b41b515a7f7018059418bab892a7b68f283d489abc4b",
+                           "NULSd6Hgj7ZoVgsPN9ybB4C1N2TbvkgLc8Z9H", "NULSd6Hgied7ym6qMEfVzZanMaa9qeqA6TZSe",
+                           10000000, 1, 1, "0000000000000000", "", 100000000);
+    auto signer = NULS::Signer(input);
```
Comment (hewigovens) [inline tests/NULS/SignerTests.cpp]:
```suggestion
    auto signer = Signer(input);
```

File: tests/NULS/SignerTests.cpp [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    Data signature = signer.sign();
+    std::string signedTransaction = hex(signature);
```
Comment (hewigovens) [inline tests/NULS/SignerTests.cpp]: is this a mainnet tx test? can we broadcast it directly via RPC?

File: swift/Tests/CoinAddressDerivationTests.swift [hunk 9]
```diff
@@ -142,6 +142,9 @@ class CoinAddressDerivationTests: XCTestCase {
                 case .qtum:
                     let expectedResult = "QhceuaTdeCZtcxmVc6yyEDEJ7Riu5gWFoF"
                     AssetCoinDerivation(coin, expectedResult, derivedAddress, address)
+                case .nuls:
+                    let expectedResult = "NsdtNvsfmPerWk4BhcapHTB3LptF8Sbe"
+                    AssetCoinDerivation(coin, expectedResult, derivedAddress, address)
```
Comment (hewigovens) [inline swift/Tests/CoinAddressDerivationTests.swift]: Any other HD wallet support NULS? could you please make sure same recovery phrase generate same address?

File: swift/Tests/Blockchains/NULSTests.swift [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        let address = NULSAddress(publicKey: pubkey)
+        let addressFromString = NULSAddress(string: "NULSd6HghWa4CN5qdxqMwYVikQxRZyj57Jn4L")!
+
+        XCTAssertEqual(address.description, addressFromString.description)
+    }
+
+    func testSign() {
+        let input = NULSSigningInput.with {
+            $0.private_key = "9ce21dad67e0f0af2599b41b515a7f7018059418bab892a7b68f283d489abc4b"
+            $0.from_address = "NULSd6Hgj7ZoVgsPN9ybB4C1N2TbvkgLc8Z9H"
```
Comment (vikmeup) [inline swift/Tests/Blockchains/NULSTests.swift]: can you rename `from_address` to `from`, and same for `to`

File: src/NULS/Address.cpp (hunk 3 above)
Comment (ccccbjcn) [reply src/NULS/Address.cpp]: How could I get the private key?

File: swift/Tests/Blockchains/NULSTests.swift (hunk 10 above)
Comment (ccccbjcn) [reply swift/Tests/Blockchains/NULSTests.swift]: Renaming will change many code. Is it mandatory?

File: tests/NULS/SignerTests.cpp (hunk 8 above)
Comment (ccccbjcn) [reply tests/NULS/SignerTests.cpp]: Yes, it is a mainnet test.

File: src/NULS/Address.cpp (hunk 3 above)
Comment (hewigovens) [reply src/NULS/Address.cpp]: use raw bytes or `PrivateKey` class

File: swift/Tests/CoinAddressDerivationTests.swift (hunk 9 above)
Comment (ccccbjcn) [reply swift/Tests/CoinAddressDerivationTests.swift]: I remove this case code.

File: src/NULS/Address.cpp (hunk 3 above)
Comment (ccccbjcn) [reply src/NULS/Address.cpp]: OK, I will fix it.

File: swift/Tests/CoinAddressDerivationTests.swift (hunk 9 above)
Comment (hewigovens) [reply swift/Tests/CoinAddressDerivationTests.swift]: why remove? just make sure we can be compatible with other wallets

File: coins.json (hunk 1 above)
Comment (ccccbjcn) [reply coins.json]: OK, nuls 2.0 has public service api link https://public1.nuls.io/

File: swift/Tests/Blockchains/NULSTests.swift (hunk 10 above)
Comment (ccccbjcn) [reply swift/Tests/Blockchains/NULSTests.swift]: I change from_address to from and to_address to to.

**Comment by ccccbjcn** [comment]
@hewigovens @vikmeup 
I have completed iOS/Android test cases. The all checks have passed.
Please check and post comment. 
Thanks.

File: src/Coin.cpp [hunk 11]
```diff
@@ -40,6 +40,9 @@
 #include "Wanchain/Address.h"
 #include "Waves/Address.h"
 #include "Zcash/TAddress.h"
+#include "NULS/Address.h"
+#include "Steem/Address.h"
+#include "IoTeX/Address.h"
```
Comment (hewigovens) [inline src/Coin.cpp]: duplicated imports

File: src/NULS/Address.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+#include <TrezorCrypto/ecdsa.h>
+
+#include "../Base58.h"
+#include "../BinaryCoding.h"
+#include "../HexCoding.h"
+
+using namespace TW;
+using namespace TW::NULS;
+
+const std::string Address::prefix("NULSd");
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: what does this line mean?

File: src/NULS/Address.cpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+bool Address::isValid(const std::string& string) {
+    if (string.empty()) {
+        return false;
+    }
+    if (string.length() <=  prefix.length()) {
+        return false;
+    }
+
+    std::string address = string.substr(prefix.length(), string.length() - prefix.length());
+    Data decoded = TW::Base58::bitcoin.decode(address);
```
Comment (hewigovens) [inline src/NULS/Address.cpp]:
```suggestion
    Data decoded = Base58::bitcoin.decode(address);
```

File: src/NULS/Address.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+        checkSum ^= decoded[i];
+    }
+
+    return decoded[23] == checkSum;
+}
+
+Address::Address(const TW::PublicKey& publicKey) {
+    // Main-Net chainID
+    bytes[0] = MainNetID;
+    bytes[1] = 0x00;
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: what's the magic value? 0x00

File: src/NULS/Address.cpp [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+    // Address Type
+    bytes[2] = addressType;
+
+    ecdsa_get_pubkeyhash(publicKey.bytes.data(), HASHER_SHA2_RIPEMD, bytes.begin() + 3);
+
+    // Calc chechsum
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: looks like we can add and resue a checksum function

File: src/NULS/Address.cpp [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+uint8_t Address::type() const {
+    return bytes[2];
+}
+
+TW::PrivateKey Address::getPrivateKey(const Data& privKey) {
+    return PrivateKey(privKey);
+}
+
+std::string Address::string() const {
+    return prefix + TW::Base58::bitcoin.encode(bytes.begin(), bytes.end());
```
Comment (hewigovens) [inline src/NULS/Address.cpp]:
```suggestion
    return prefix + Base58::bitcoin.encode(bytes.begin(), bytes.end());
```

File: src/NULS/Address.h [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    static const std::string prefix;
+
+    /// NULS address type
+    static const byte addressType = 0x01;
+
+    /// Determines whether a string makes a valid address.
+    static bool isValid(const std::string& string);
+
+    /// get NULS private key
+    static PrivateKey getPrivateKey(const Data& privKey);
```
Comment (hewigovens) [inline src/NULS/Address.h]: looks trivial and can be removed, use `PrivateKey(data)` directly

File: src/NULS/Address.h [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+
+    /// Initializes an address from a string representation.
+    explicit Address(const std::string& string);
+
+    /// Initializes an address from a public key.
+    explicit Address(const PublicKey& publicKey);
+
+    /// Initializes an address with a collection of bytes.
+    explicit Address(const std::vector<uint8_t>& data) : TW::Base58Address<24>(data) {}
```
Comment (hewigovens) [inline src/NULS/Address.h]:
```suggestion
    explicit Address(const Data& data) : TW::Base58Address<24>(data) {}
```

File: src/NULS/BinaryCoding.h [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+#pragma once
+#include "../uint256.h"
+#include "../BinaryCoding.h"
+#include "../proto/NULS.pb.h"
+#include "../HexCoding.h"
+
+using namespace TW;
+using namespace TW::NULS;
+
+static inline void serializerRemark(std::string& remark, std::vector<uint8_t>& data) {
```
Comment (hewigovens) [inline src/NULS/BinaryCoding.h]:
```suggestion
static inline void serializerRemark(std::string& remark, Data& data) {
```

File: src/NULS/BinaryCoding.h [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    encodeVarInt(remark.length(), data);
+    std::copy(remark.begin(), remark.end(), std::back_inserter(data));
+}
+
+static inline void serializerInput(const Proto::TransactionCoinFrom& input,
+    std::vector<uint8_t>& data) {
```
Comment (hewigovens) [inline src/NULS/BinaryCoding.h]:
```suggestion
    Data& data) {
```

File: src/NULS/BinaryCoding.h [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+    encodeVarInt(1, data);  //there is one coinFrom
+    const auto& fromAddress = input.from_address();
+    if (!NULS::Address::isValid(fromAddress)) {
+        throw std::invalid_argument("Invalid address");
+    }
+    const auto& addr = NULS::Address(fromAddress);
+    encodeVarInt(addr.bytes.size() - 1, data);
+    std::copy(addr.bytes.begin(), addr.bytes.end() - 1, std::back_inserter(data));
+    encode16LE((uint16_t)input.assets_chainid(), data);
```
Comment (hewigovens) [inline src/NULS/BinaryCoding.h]: use `static_cast`

File: src/NULS/BinaryCoding.h [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+    encode16LE((uint16_t)input.assets_id(), data);
+    std::copy(input.id_amount().begin(), input.id_amount().end(), std::back_inserter(data));
+    Data nonce = parse_hex(input.nonce());
+    encodeVarInt(nonce.size(), data);
+    append(data, nonce);
+    data.push_back(static_cast<uint8_t>(input.locked()));
+}
+
+static inline void serializerOutput(const Proto::TransactionCoinTo& output,
+    std::vector<uint8_t>& data) {
```
Comment (hewigovens) [inline src/NULS/BinaryCoding.h]:
```suggestion
    Data& data) {
```

File: src/NULS/BinaryCoding.h [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+
+    const auto& toAddress = output.to_address();
+    if (!NULS::Address::isValid(toAddress)) {
+        throw std::invalid_argument("Invalid address");
+    }
+    const auto& addr = NULS::Address(toAddress);
+    encodeVarInt(addr.bytes.size() - 1, data);
+    std::copy(addr.bytes.begin(), addr.bytes.end() - 1, std::back_inserter(data));
+    encode16LE((uint16_t)output.assets_chainid(), data);
+    encode16LE((uint16_t)output.assets_id(), data);
```
Comment (hewigovens) [inline src/NULS/BinaryCoding.h]: static_cast

File: src/NULS/Signer.cpp [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+    Proto::TransactionCoinTo coinTo;
+    coinTo.set_to_address(input.to());
+    coinTo.set_id_amount(input.amount());
+    coinTo.set_assets_chainid(input.chain_id());
+    coinTo.set_assets_id(input.idassets_id());
+    coinTo.set_lock_time(0);
+    *tx.mutable_output() = coinTo;
+
+    tx.set_remark(input.remark());
+    tx.set_type(2);
```
Comment (hewigovens) [inline src/NULS/Signer.cpp]: what's this magic value?

File: src/NULS/Signer.cpp [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+
+    return dataRet;
+}
+
+Data Signer::sign(Proto::SigningInput& input) const {
+    Signer signer = Signer(input);
+    return signer.sign();
+}
+
+uint32_t Signer::CalculatorTransactionSize(uint32_t inputCount, uint32_t outputCount, uint32_t remarkSize) const {
```
Comment (hewigovens) [inline src/NULS/Signer.cpp]: Is uint32_t  enough? any chance to overflow?

Comment (ccccbjcn) [reply src/NULS/Signer.cpp]: no, it is enough.

File: src/NULS/Address.cpp (hunk 12 above)
Comment (ccccbjcn) [reply src/NULS/Address.cpp]: Nuls 2.0 mainnet address prefix is "NULSd".

File: src/NULS/Address.cpp (hunk 14 above)
Comment (ccccbjcn) [reply src/NULS/Address.cpp]: mainnet id is 0x0001.

File: src/NULS/Signer.cpp (hunk 24 above)
Comment (ccccbjcn) [reply src/NULS/Signer.cpp]: this is nuls 2.0 transaction type. I will define macro.

File: src/NULS/Address.cpp (hunk 14 above)
Comment (hewigovens) [reply src/NULS/Address.cpp]: can you  use bytes then

File: src/NULS/Address.cpp (hunk 15 above)
Comment (ccccbjcn) [reply src/NULS/Address.cpp]: I did not look up the checksum function.
Your mean I wrap one in the NULS code?

File: src/proto/NULS.proto [hunk 26]
```diff
@@ -0,0 +1,66 @@
+syntax = "proto3";
+
+package TW.NULS.Proto;
+option java_package = "wallet.core.jni.proto";
+
+message TransactionCoinFrom {
+    string from_address = 1;
+    uint32 assets_chainid = 2;
+    uint32 assets_id = 3;
+    //tranaction out amount (128-bit number)
```
Comment (hewigovens) [inline src/proto/NULS.proto]:
```suggestion
    //tranaction out amount (256-bit number)
```

File: src/proto/NULS.proto [hunk 27]
```diff
@@ -0,0 +1,66 @@
...
+    bytes nonce = 5;
+    //lock status: 1 locked; 0 unlocked
+    uint32 locked = 6;
+}
+
+message TransactionCoinTo {
+    string to_address = 1;
+    uint32 assets_chainid = 2;
+    uint32 assets_id = 3;
+    // tranaction amount (128-bit number)
```
Comment (hewigovens) [inline src/proto/NULS.proto]: loaded in uint256_t
```suggestion
    // tranaction amount (256-bit number)
```

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nuls/TestNULSSigner.kt [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+    @Test
+    fun NULSTransactionSigning() { 
+        val signingInput = NULS.SigningInput.newBuilder()
+            .setPrivateKey(ByteString.copyFrom(Numeric.hexStringToByteArray("9ce21dad67e0f0af2599b41b515a7f7018059418bab892a7b68f283d489abc4b")))
+            .setFrom("NULSd6Hgj7ZoVgsPN9ybB4C1N2TbvkgLc8Z9H")
+            .setTo("NULSd6Hgied7ym6qMEfVzZanMaa9qeqA6TZSe")
+            .setAmount(ByteString.copyFrom("0x989680".toHexByteArray()))
+            .setChainId(1)
+            .setIdassetsId(1)
+            .setNonce(ByteString.copyFrom("0000000000000000".toByteArray()))
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nuls/TestNULSSigner.kt]: how to get the nonce and convert it to bytes?

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nuls/TestNULSSigner.kt [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+            .setRemark("")
+            .setBalance(ByteString.copyFrom("0x05f5e100".toHexByteArray()))
+            .setTimestamp(0x5d8885f8)
```
Comment (hewigovens) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nuls/TestNULSSigner.kt]: seconds or milliseconds? can you add comments to proto file?

File: src/Coin.cpp [hunk 30]
```diff
@@ -40,6 +41,7 @@
 #include "Wanchain/Address.h"
 #include "Waves/Address.h"
 #include "Zcash/TAddress.h"
+#include "IoTeX/Address.h"
```
Comment (hewigovens) [inline src/Coin.cpp]: duplicated import again, why don't you look at the diff by yourself?

File: src/NULS/Address.cpp (hunk 15 above)
Comment (hewigovens) [reply src/NULS/Address.cpp]: yes, wrap a checksum function

File: src/NULS/Address.cpp [hunk 31]
```diff
@@ -0,0 +1,79 @@
...
+    for (int i = 0; i < 23; ++i) {
+        checkSum ^= decoded[i];
+    }
+
+    return decoded[23] == checkSum;
+}
+
+Address::Address(const TW::PublicKey& publicKey) {
+    // Main-Net chainID
+    bytes[0] = MainNetID;
```
Comment (hewigovens) [inline src/NULS/Address.cpp]: something like: mainetId = {0x01, 0x00}

File: src/proto/NULS.proto [hunk 32]
```diff
@@ -0,0 +1,66 @@
+syntax = "proto3";
+
+package TW.NULS.Proto;
+option java_package = "wallet.core.jni.proto";
+
+message TransactionCoinFrom {
```
Comment (hewigovens) [inline src/proto/NULS.proto]: Are the naming `CoinFrom` and `CoinTo` also used in NULS code base? can we rename to Input/Output?

File: src/proto/NULS.proto [hunk 33]
```diff
@@ -0,0 +1,66 @@
...
+    string to = 3;
+    bytes  amount = 4;
+    uint32 chain_id = 5;
+    uint32 idassets_id = 6;
+    //The last 8 bytes of latest transaction hash 
+    bytes nonce = 7;
+    string remark = 8;
+    // Account balance 
+    bytes  balance = 9;
+    uint32 timestamp = 10;
```
Comment (hewigovens) [inline src/proto/NULS.proto]: seconds or milliseconds?

File: src/NULS/Address.cpp (hunk 14 above)
Comment (ccccbjcn) [reply src/NULS/Address.cpp]: @hewigovens  what is your mean?

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nuls/TestNULSSigner.kt (hunk 28 above)
Comment (ccccbjcn) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nuls/TestNULSSigner.kt]: nonce : //byte[8] 
transaction sequence number, the last 8 bytes of the hash of the previous transaction

File: src/Coin.cpp (hunk 30 above)
Comment (ccccbjcn) [reply src/Coin.cpp]: it maybe git merge issue.

File: src/proto/NULS.proto (hunk 32 above)
Comment (ccccbjcn) [reply src/proto/NULS.proto]: They are used by NULS code only and their names are same with NULS definition.
I think they need not to be changed.

File: src/proto/NULS.proto (hunk 33 above)
Comment (ccccbjcn) [reply src/proto/NULS.proto]: second.
I will comment for this field.

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nuls/TestNULSSigner.kt (hunk 29 above)
Comment (ccccbjcn) [reply android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/nuls/TestNULSSigner.kt]: second, I will add comment.

**Comment by ccccbjcn** [comment]
> Can you fix all the comments? Thanks

I fix your comment.

**Comment by ccccbjcn** [comment]
@leoneparise @hewigovens 
Please review again. Thanks.

**Comment by ccccbjcn** [comment]
Thank @hewigovens  @leoneparise approvals.

@hewigovens This PR all checks have passed and proved by you and leoneparise. 
How the PR will be merged into to TW repository?

**Comment by ccccbjcn** [comment]
@hewigovens  Thanks.

**Pull request merged**